wasm_node_api/wasm_node_api.cc

```diff
@@ -0,0 +1,886 @@
+    return w::ValKind::F32;
+  } else if constexpr (std::is_same<kind<T>, double>::value) {
+    return w::ValKind::F64;
+  #if __WORDSIZE == 64
+  } else if constexpr (is_native_v<T> && std::is_pointer<kind<T>>::value) {
+    return w::ValKind::I64;
+  #endif
+  } else {
+    return w::ValKind::I32;
+  }
+}
+
+template<w::ValKind... Args>
+constexpr auto exportVoid(callback cb) -> w::Func {
+  return w::Func(
+    w::FuncType({Args...}, {}),
+    cb
+  );
+};
+
+template<w::ValKind Return, w::ValKind... Args>
+constexpr auto exportFn(callback cb) -> w::Func {
+  return w::Func(
+    w::FuncType({Args...}, {Return}),
+    cb
+  );
+};
+
+template<typename Return, typename... Args>
+constexpr auto exportNative(callback cb) -> w::Func {
+  if constexpr(std::is_void<kind<Return>>::value) {
+    return exportVoid<wasmType<Args>()...>(cb);
+  } else {
+    return exportFn<wasmType<Return>(), wasmType<Args>()...>(cb);
+  }
+}
+
+template<typename Arg>
+constexpr Arg napiCast(const w::Val& v) {
+  if constexpr(std::is_same<Arg, float>::value) {
+    return (Arg) v.f32();
+  } else if constexpr(std::is_same<Arg, double>::value) {
+    return (Arg) v.f64();
+  } else {
+    return (Arg) v.i32();
+  }
+}
+
+template<typename Result>
+constexpr kind<Result> napiUnwrap(const w::Context* ctx, const w::Val& v) {
+  if constexpr(is_wasm_v<Result>) {
+    if constexpr(/*std::is_same<kind<Result>,
+                              v8::MaybeLocal<v8::Function>>::value*/
+                 is_function_pointer_v<kind<Result>>) {
+      return (kind<Result>) v.i32();/*(ctx->table->get(v.i32()));*/
+    } else {
+      return (kind<Result>) (&ctx->memory->data()[(uint32_t) v.i32()]);
+    }
+  #if __WORDSIZE == 64
+  } else if constexpr(std::is_pointer<kind<Result>>::value) { // Native pointer
+    return (kind<Result>) v.i64();
+  #endif
+  } else {
+    return napiCast<kind<Result>>(v);
+  }
+};
+
+template<typename Result>
+constexpr w::Val napiWrap(kind<Result> v) {
+  if constexpr(std::is_same<kind<Result>, float>::value) {
+    return w::Val((float) v);
+  } else if constexpr (std::is_same<kind<Result>, double>::value) {
+    return w::Val((double) v);
+  #if __WORDSIZE == 64
+  } else if constexpr(std::is_pointer<kind<Result>>::value) { // Native pointer
+      return w::Val(static_cast<int64_t>((size_t) v));
+  #endif
+  } else {
+    return w::Val(static_cast<int32_t>((size_t) v));
+  }
+}
+
+template<typename Result, typename... Args>
+struct napiApply {
+  template<size_t... Indices>
+  static constexpr auto apply(
+    kind<Result> (*fn)(kind<Args>...),
+    const w::Context* ctx,
+    const w::Val args[],
+    std::index_sequence<Indices...>
+  ) -> kind<Result> {
+    if constexpr(std::is_void<kind<Result>>::value) {
+      fn(napiUnwrap<Args>(ctx, args[Indices])...);
+    } else {
+      return fn(napiUnwrap<Args>(ctx, args[Indices])...);
+    }
+  }
+};
+
+template<typename Result, typename... Args>
+struct napiCall {
```

Comment:
`napiCall` is an abstraction over `napiApply` that performs the entire call operation into a N-API function. By taking the function pointer as a template parameter, this generates a static call wrapper for the N-API function. It includes some of the boilerplate for making `napiApply` calls and also handles the return values from `napiApply` and places them into the `results` array if necessary.